elastic/kibana
# Avoid loading all plugins on anonymous pages

Related to #121299.

### Overview

The rendering service returns the list of UI plugins to the browser in `kbn-injected-metadata`.
This PR adds an "enableForAnonymousPages" plugin manifest property.
This boolean property specifies whether the plugin is included in that list when the page is anonymous. Now, any plugins that do not explicitly specify this property, and/or are not a required dependency of _another_ plugin that does specify it, are not enabled on anonymous pages (login page, status page when configured, etc.)

This PR also limits the plugin _page load bundles_ that are loaded on anonymous pages.

### Testing

1. Open the Kibana login page in your browser.
2. Using your browser's developer tools, find the source for `bootstrap.js` and add a breakpoint to the line that bootstraps the plugins.
3. Refresh the page and wait for it to catch on the breakpoint.
4. Execute this line in the console to parse the injected metadata from the DOM:
   ```js
   JSON.parse(document.querySelector('kbn-injected-metadata').getAttribute('data'));
   ```
5. Observe that only three plugins are included in the injected metadata (`licensing`, `features`, and `security`).
6. Resume execution.
7. Log in and wait for it to catch on the breakpoint again.
8. Parse the injected metadata from the DOM again.
9. Observe that now all plugins are included in the injected metadata.

You can also test by configuring Kibana to allow anonymous access to the status page (by adding `status.allowAnonymous: true` to your kibana.yml) and navigating to the `/status` page; you'll see that while viewing this as anonymously (without an authentication cookie), only the three aforementioned plugins are loaded.

### Notes

"Anonymous pages" are rendered using the `renderAnonymousCoreApp` function (which calls the rendering service with `isAnonymousPage: true`), while "Authenticated pages" are rendered using the `renderCoreApp` function (which calls the rendering service with `isAnonymousPage: false`).

### Screenshots

Injected metadata on the login page:
![Screen Shot 2022-04-06 at 3 56 06 PM](https://user-images.githubusercontent.com/5295965/162060804-3410ab27-bc63-4750-b734-e2c6e8151bf8.png)

Injected metadata on the home page:
![Screen Shot 2022-04-05 at 11 21 51 PM](https://user-images.githubusercontent.com/5295965/161892149-b69ea7d4-3818-4002-909e-fa3c65cdd98c.png)

JS bundles that are loaded on the login page:

![image](https://user-images.githubusercontent.com/5295965/162069187-0ce5fa4e-1819-496b-bc1e-73158c75a9bb.png)

JS bundles that are loaded on the home page:

![image](https://user-images.githubusercontent.com/5295965/162069289-7c7ddc06-24dd-4e07-a73f-e2837b94172c.png)
